File: src/documents.ts
import { File } from 'expo-file-system';
import * as Print from 'expo-print';
import * as Sharing from 'expo-sharing';
import type { BusinessProfile, Payment, Quote } from './domain';
import { formatFolio, money, normalizeBrandColor, paymentSummary } from './domain';

const escapeHtml = (value: string) => value.replace(/[&<>'"]/g, char => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', "'": '&#39;', '"': '&quot;' })[char] ?? char);
const date = (value: string) => new Date(/^\d{4}-\d{2}-\d{2}$/.test(value) ? `${value}T12:00:00Z` : value).toLocaleDateString('es-MX', { day: 'numeric', month: 'long', year: 'numeric' });

async function logoDataUri(profile: BusinessProfile, isPro: boolean) {
  if (!isPro || !profile.logoUri) return '';
  const file = new File(profile.logoUri); if (!file.exists) return '';
  const base64 = await file.base64();
  const mimeType = base64.startsWith('/9j/') ? 'image/jpeg' : base64.startsWith('iVBOR') ? 'image/png' : '';
  return mimeType ? `data:${mimeType};base64,${base64}` : '';
}

function shell(profile: BusinessProfile, isPro: boolean, body: string, logo: string) {
  const accent = isPro ? normalizeBrandColor(profile.brandColor) : '#0E5E4A';
  return `<!doctype html><html><head><meta charset="utf-8"><style>
    @page{margin:24px}body{font-family:Arial,sans-serif;color:#17221d;font-size:12px;margin:0}.header{display:flex;justify-content:space-between;align-items:flex-start;border-bottom:3px solid ${accent};padding-bottom:16px;margin-bottom:22px}.logo{max-width:100px;max-height:70px}.business{font-size:23px;font-weight:800;color:${accent}}.muted{color:#69746f}.meta{line-height:1.6;text-align:right}h1{font-size:18px;margin:0 0 5px}table{width:100%;border-collapse:collapse;margin:20px 0}th{background:#f1f5f3;color:${accent};font-size:10px;text-transform:uppercase}td,th{padding:10px 7px;border-bottom:1px solid #dfe5e2;text-align:left}.number{text-align:right}.total{font-size:22px;font-weight:800;text-align:right;color:${accent}}.box{border:1px solid #dfe5e2;border-radius:8px;padding:12px;margin-top:14px}.legal{margin-top:36px;border-top:1px solid #ddd;padding-top:10px;font-size:10px;color:#777}.jale{text-align:center;margin-top:14px;color:#777;font-size:10px}</style></head><body>
    <div class="header"><div>${logo ? `<img class="logo" src="${logo}">` : ''}<div class="business">${escapeHtml(profile.name)}</div><div class="muted">${escapeHtml(profile.trade)}${profile.phone ? ` · ${escapeHtml(profile.phone)}` : ''}</div></div>${body.split('<!--META-->')[0]}</div>${body.split('<!--META-->')[1]}
    <div class="legal">Este documento no es un comprobante fiscal.</div>${isPro ? '' : '<div class="jale">Creado con Jale</div>'}</body></html>`;
}

export async function quoteHtml(profile: BusinessProfile, quote: Quote, isPro: boolean) {
  const rows = quote.lines.map(line => `<tr><td>${escapeHtml(line.concept)}</td><td class="number">${line.quantity}</td><td class="number">${money(line.unitPriceCents)}</td><td class="number">${money(Math.round(line.quantity * line.unitPriceCents))}</td></tr>`).join('');
  const meta = `<div class="meta"><h1>COTIZACIÓN</h1><b>${formatFolio('COT', quote.quoteNumber)}</b><br>${date(quote.issuedAt)}${quote.validUntil ? `<br>Válida hasta ${date(quote.validUntil)}` : ''}</div>`;
  const content = `<section><b>Cliente</b><br>${escapeHtml(quote.clientName)}</section><table><thead><tr><th>Concepto</th><th class="number">Cant.</th><th class="number">Precio</th><th class="number">Importe</th></tr></thead><tbody>${rows}</tbody></table><div class="total">Total: ${money(quote.totalCents)}</div>${quote.notes ? `<div class="box"><b>Notas</b><br>${escapeHtml(quote.notes)}</div>` : ''}`;
  return shell(profile, isPro, `${meta}<!--META-->${content}`, await logoDataUri(profile, isPro));
}

export async function receiptHtml(profile: BusinessProfile, quote: Quote, payment: Payment, isPro: boolean) {
  const summary = paymentSummary(quote.totalCents, quote.payments);
  const meta = `<div class="meta"><h1>RECIBO DE PAGO</h1><b>${formatFolio('REC', payment.receiptNumber)}</b><br>${date(payment.paidAt)}</div>`;
  const method = payment.method === 'cash' ? 'Efectivo' : payment.method === 'transfer' ? 'Transferencia' : 'Otro';
  const content = `<section><b>Recibimos de</b><br>${escapeHtml(quote.clientName)}</section><div class="box"><b>Por concepto de</b><br>Pago de ${formatFolio('COT', quote.quoteNumber)}${payment.notes ? ` · ${escapeHtml(payment.notes)}` : ''}</div><p>Forma de pago: <b>${method}</b></p><div class="total">Recibido: ${money(payment.amountCents)}</div><div class="box">Total de la cotización: ${money(quote.totalCents)}<br>Pagado acumulado: ${money(summary.paidCents)}<br><b>Saldo pendiente: ${money(summary.balanceCents)}</b></div>`;
  return shell(profile, isPro, `${meta}<!--META-->${content}`, await logoDataUri(profile, isPro));
}

async function shareHtml(html: string, title: string) {
  const { uri } = await Print.printToFileAsync({ html });
  if (!(await Sharing.isAvailableAsync())) throw new Error('La función de compartir no está disponible.');
  await Sharing.shareAsync(uri, { mimeType: 'application/pdf', UTI: 'com.adobe.pdf', dialogTitle: title });
  return uri;
}

export async function shareQuotePdf(profile: BusinessProfile, quote: Quote, isPro: boolean) { return shareHtml(await quoteHtml(profile, quote, isPro), `Compartir ${formatFolio('COT', quote.quoteNumber)}`); }
export async function shareReceiptPdf(profile: BusinessProfile, quote: Quote, payment: Payment, isPro: boolean) { return shareHtml(await receiptHtml(profile, quote, payment, isPro), `Compartir ${formatFolio('REC', payment.receiptNumber)}`); }

File: lib/documents.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';
import 'domain.dart';

PdfColor _color(String hex) =>
    PdfColor.fromInt(0xFF000000 | int.parse(hex.substring(1), radix: 16));
String _date(String value) {
  final parsed = DateTime.tryParse(value);
  return parsed == null
      ? value
      : '${parsed.day}/${parsed.month}/${parsed.year}';
}

String _method(PaymentMethod method) => switch (method) {
  PaymentMethod.cash => 'Efectivo',
  PaymentMethod.transfer => 'Transferencia',
  PaymentMethod.other => 'Otro',
};

Future<Uint8List> _document({
  required BusinessProfile profile,
  required Quote quote,
  Payment? payment,
  required bool isPro,
}) async {
  final pdf = pw.Document();
  final accent = _color(normalizeBrandColor(profile.brandColor));
  pw.MemoryImage? logo;
  if (profile.logoUri != null && profile.logoUri!.isNotEmpty) {
    try {
      logo = pw.MemoryImage(await File(profile.logoUri!).readAsBytes());
    } catch (_) {
      logo = null;
    }
  }
  final summary = paymentSummary(quote.totalCents, quote.payments);
  pdf.addPage(
    pw.MultiPage(
      pageFormat: PdfPageFormat.letter,
      margin: const pw.EdgeInsets.all(28),
      footer: (context) => pw.Column(
        children: [
          pw.Divider(),
          pw.Row(
            mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
            children: [
              pw.Text(
                'Este documento no es un comprobante fiscal.',
                style: const pw.TextStyle(fontSize: 9, color: PdfColors.grey),
              ),
              pw.Text(
                'Página ${context.pageNumber} de ${context.pagesCount}',
                style: const pw.TextStyle(fontSize: 9, color: PdfColors.grey),
              ),
            ],
          ),
          if (!isPro)
            pw.Text(
              'Creado con Jale',
              style: const pw.TextStyle(fontSize: 9, color: PdfColors.grey),
            ),
        ],
      ),
      build: (_) {
        final title = payment == null ? 'COTIZACIÓN' : 'RECIBO DE PAGO';
        final folio = formatFolio(
          payment == null ? 'COT' : 'REC',
          payment == null ? quote.quoteNumber : payment.receiptNumber,
        );
        final children = <pw.Widget>[
          pw.Container(
            padding: const pw.EdgeInsets.only(bottom: 12),
            decoration: pw.BoxDecoration(
              border: pw.Border(bottom: pw.BorderSide(color: accent, width: 3)),
            ),
            child: pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                if (logo != null) ...[
                  pw.ClipRRect(
                    horizontalRadius: 8,
                    verticalRadius: 8,
                    child: pw.Image(logo, width: 54, height: 54),
                  ),
                  pw.SizedBox(width: 10),
                ],
                pw.Expanded(
                  child: pw.Column(
                    crossAxisAlignment: pw.CrossAxisAlignment.start,
                    children: [
                      pw.Text(
                        profile.name,
                        style: pw.TextStyle(
                          fontSize: 22,
                          fontWeight: pw.FontWeight.bold,
                          color: accent,
                        ),
                      ),
                      if (profile.trade.isNotEmpty)
                        pw.Text(
                          '${profile.trade}${profile.phone.isEmpty ? '' : ' · ${profile.phone}'}',
                          style: const pw.TextStyle(color: PdfColors.grey),
                        ),
                    ],
                  ),
                ),
                pw.SizedBox(width: 10),
                pw.Column(
                  crossAxisAlignment: pw.CrossAxisAlignment.end,
                  children: [
                    pw.Text(
                      title,
                      style: pw.TextStyle(
                        fontWeight: pw.FontWeight.bold,
                        color: accent,
                      ),
                    ),
                    pw.Text(
                      folio,
                      style: pw.TextStyle(fontWeight: pw.FontWeight.bold),
                    ),
                    pw.Text(_date(payment?.paidAt ?? quote.issuedAt)),
                  ],
                ),
              ],
            ),
          ),
          pw.SizedBox(height: 22),
        ];
        if (payment == null) {
          children.addAll([
            pw.Text(
              'Cliente',
              style: pw.TextStyle(fontWeight: pw.FontWeight.bold),
            ),
            pw.Text(quote.clientName),
            if (quote.clientPhone.isNotEmpty) pw.Text(quote.clientPhone),
            if (quote.clientAddress.isNotEmpty) pw.Text(quote.clientAddress),
            pw.SizedBox(height: 14),
          ]);
          final rows = <pw.TableRow>[
            pw.TableRow(
              decoration: const pw.BoxDecoration(color: PdfColors.grey100),
              children: ['Concepto', 'Cant.', 'Precio', 'Importe']
                  .map(
                    (item) => pw.Padding(
                      padding: const pw.EdgeInsets.all(7),
                      child: pw.Text(
                        item,
                        style: pw.TextStyle(
                          fontWeight: pw.FontWeight.bold,
                          color: accent,
                        ),
                      ),
                    ),
                  )
                  .toList(),
            ),
          ];
          rows.addAll(
            quote.lines.map(
              (line) => pw.TableRow(
                children:
                    [
                          line.concept,
                          '${line.quantity}',
                          money(line.unitPriceCents),
                          money((line.quantity * line.unitPriceCents).round()),
                        ]
                        .map(
                          (item) => pw.Padding(
                            padding: const pw.EdgeInsets.all(7),
                            child: pw.Text(item),
                          ),
                        )
                        .toList(),
              ),
            ),
          );
          children.add(
            pw.Table(
              border: pw.TableBorder(
                horizontalInside: pw.BorderSide(color: PdfColors.grey200),
              ),
              children: rows,
            ),
          );
          children.addAll([
            pw.SizedBox(height: 16),
            pw.Align(
              alignment: pw.Alignment.centerRight,
              child: pw.Text(
                'Total: ${money(quote.totalCents)}',
                style: pw.TextStyle(
                  fontSize: 18,
                  fontWeight: pw.FontWeight.bold,
                  color: accent,
                ),
              ),
            ),
          ]);
          if (quote.validUntil != null) {
            children.add(pw.Text('Válida hasta ${_date(quote.validUntil!)}'));
          }
          if (quote.notes.isNotEmpty) {
            children.add(
              pw.Container(
                margin: const pw.EdgeInsets.only(top: 14),
                padding: const pw.EdgeInsets.all(10),
                decoration: pw.BoxDecoration(
                  border: pw.Border.all(color: PdfColors.grey300),
                ),
                child: pw.Text('Notas\n${quote.notes}'),
              ),
            );
          }
        } else {
          children.addAll([
            pw.Text(
              'Recibimos de',
              style: pw.TextStyle(fontWeight: pw.FontWeight.bold),
            ),
            pw.Text(quote.clientName),
            pw.SizedBox(height: 14),
            pw.Container(
              width: double.infinity,
              padding: const pw.EdgeInsets.all(12),
              decoration: pw.BoxDecoration(
                border: pw.Border.all(color: PdfColors.grey300),
              ),
              child: pw.Text(
                'Pago de ${formatFolio('COT', quote.quoteNumber)}${payment.notes.isEmpty ? '' : ' · ${payment.notes}'}',
              ),
            ),
            pw.SizedBox(height: 12),
            pw.Text('Forma de pago: ${_method(payment.method)}'),
            pw.SizedBox(height: 18),
            pw.Align(
              alignment: pw.Alignment.centerRight,
              child: pw.Text(
                'Recibido: ${money(payment.amountCents)}',
                style: pw.TextStyle(
                  fontSize: 18,
                  fontWeight: pw.FontWeight.bold,
                  color: accent,
                ),
              ),
            ),
            pw.SizedBox(height: 12),
            pw.Text(
              'Total de la cotización: ${money(quote.totalCents)}\nPagado acumulado: ${money(summary.paidCents)}\nSaldo pendiente: ${money(summary.balanceCents)}',
            ),
          ]);
        }
        return children;
      },
    ),
  );
  return pdf.save();
}

Future<void> shareQuotePdf(
  BusinessProfile profile,
  Quote quote,
  bool isPro,
) async => Printing.sharePdf(
  bytes: await quotePdfBytes(profile, quote, isPro),
  filename: '${formatFolio('COT', quote.quoteNumber)}.pdf',
);

Future<Uint8List> quotePdfBytes(
  BusinessProfile profile,
  Quote quote,
  bool isPro,
) => _document(profile: profile, quote: quote, isPro: isPro);

Future<void> shareReceiptPdf(
  BusinessProfile profile,
  Quote quote,
  Payment payment,
  bool isPro,
) async => Printing.sharePdf(
  bytes: await _document(
    profile: profile,
    quote: quote,
    payment: payment,
    isPro: isPro,
  ),
  filename: '${formatFolio('REC', payment.receiptNumber)}.pdf',
);
